L1Trigger/L1THGCal/interface/be_algorithms/HGCalTowerMap2DImpl.h

```diff
@@ -0,0 +1,54 @@
+#ifndef __L1Trigger_L1THGCal_HGCalTowerMap2DImpl_h__
+#define __L1Trigger_L1THGCal_HGCalTowerMap2DImpl_h__
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
+#include "DataFormats/L1THGCal/interface/HGCalCluster.h"
+#include "L1Trigger/L1THGCal/interface/HGCalTriggerTools.h"
+
+
+class HGCalTowerMap2DImpl{
+
+ public:
+
+  HGCalTowerMap2DImpl( const edm::ParameterSet &conf); 
+  
+  void eventSetup(const edm::EventSetup& es) 
+    {
+        triggerTools_.setEventSetup(es);
+    }
+
+  void buildTowerMapClusters( const std::vector<edm::Ptr<l1t::HGCalTriggerCell>> & triggerCellsPtrs,
+			      l1t::HGCalClusterBxCollection & clusters
+			      );
+
+  void calibratePt( l1t::HGCalCluster & cluster );  
```

Comment:
Hmm, how would the TC calibration actually work?
It would not only compute the MIPT from the hwPT in ADC counts but also the physical pT value, right?
That means that each time the TC calibration is modified, the layer calibration for the clusters would in principle need to be updated (which from experience of the current system is not very optimal).
Would it make sense to only keep the hwPT->MIPT calibration in the TC calibration and to have two different layer calibrations MIPT->physical pT for clusters and towers ?

Reply:
What is currently implemented for the TC is a calibration taking into account the absorbers (dE/dX weights), applied on top of the mip calibration. In principle this calibration doesn't change, unless we modify the layers layout of the detector.

For practical reasons, we need to have fully calibrated TCs, for the front-end studies. So the implementation of the TC calibration should be kept.

But I understand that it can be interesting to also have a dedicated calibration for the towers, in case we would like to apply some calibration different from the dE/dX calibration without touching the TCs.
But in that case it would be good to avoid cloning the calibration code in multiple places. It would be nice if it can be implemented only once in the calibration class, and then can be called in the clustering and "towering".

Reply:
Makes sense
Working on it